### PR TITLE
util: add buffered metric client

### PR DESCRIPTION
**Pull request by overvenus**
A buffered metric sink with a pseudo timeout mechanism.

Benchmark([link](https://gist.github.com/overvenus/31b3adba8d60fbc81325a958a0e9fc9d)):

```
➜  bench_metric git:(master) ✗ cargo bench bench_single                   
     Running target/release/bench_metric-b6de8d742c9d8b0f

running 3 tests
test tests::bench_single_thread_block_udp    ... bench:       3,492 ns/iter (+/- 137)
test tests::bench_single_thread_buffer_udp   ... bench:         135 ns/iter (+/- 4)
test tests::bench_single_thread_nonblock_udp ... bench:       3,487 ns/iter (+/- 530)

test result: ok. 0 passed; 0 failed; 0 ignored; 3 measured
```

3 threads send 15000 metrics concurrently.

```
➜  bench_metric git:(master) ✗ cargo bench bench_multi_thread_block_udp
     Running target/release/bench_metric-b6de8d742c9d8b0f

running 1 test
test tests::bench_multi_thread_block_udp     ... bench:  37,487,056 ns/iter (+/- 3,148,835)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured

➜  bench_metric git:(master) ✗ cargo bench bench_multi_thread_nonblock_udp
     Running target/release/bench_metric-b6de8d742c9d8b0f

running 1 test
test tests::bench_multi_thread_nonblock_udp  ... bench:  37,389,796 ns/iter (+/- 3,202,220)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured

➜  bench_metric git:(master) ✗ cargo bench bench_multi_thread_buffer_udp  
     Running target/release/bench_metric-b6de8d742c9d8b0f

running 1 test
test tests::bench_multi_thread_buffer_udp    ... bench:   4,425,618 ns/iter (+/- 230,592)

test result: ok. 0 passed; 0 failed; 0 ignored; 1 measured
```

### CODE REVIEW

**Comment by overvenus** [comment]
#668 

### One thread sends 1000000 metrics

`BufferedUdpMetricSink`:

```
time ./target/release/profile-metric
0.10s user 0.04s system 99% cpu 0.141 total
```

`NoneblockUdpMetricSink`:

```
time ./target/release/profile-metric
0.11s user 3.40s system 99% cpu 3.514 total
```

File: src/util/metric/mod.rs [hunk 1]
```diff
@@ -87,15 +94,40 @@ impl NonblockUdpMetricSink {
...
+            *last_flush_time = Instant::now();
+        }
+
+        buffer.extend_from_slice(metric.as_bytes());
+        buffer.push('\n' as u8);
+        Ok(metric.len())
+    }
+
+    fn flush(&self, buffer: &mut Vec<u8>) {
+        let _ = self.socket.send_to(buffer.as_slice(), &self.sink_addr);
```
Comment (siddontang) [inline src/util/metric/mod.rs]: log error here.

File: src/util/metric/mod.rs [hunk 2]
```diff
@@ -87,15 +94,40 @@ impl NonblockUdpMetricSink {
         // Moves this UDP stream into nonblocking mode.
         try!(socket.set_nonblocking(true));
 
-        Ok(NonblockUdpMetricSink {
+        Ok(BufferedUdpMetricSink {
             sink_addr: addr,
             socket: socket,
+
+            buffer: Mutex::new(Vec::with_capacity(1024)),
```
Comment (siddontang) [inline src/util/metric/mod.rs]: use a const variable.

**Comment by overvenus** [comment]
PTAL @BusyJay @ngaut

**Comment by siddontang** [comment]
Cool, LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**